internal/nix/shell.go

```diff
@@ -70,14 +71,63 @@ type Shell struct {
 
 type ShellOption func(*Shell)
 
-// DetectShell attempts to determine the user's default shell.
-func DetectShell(opts ...ShellOption) (*Shell, error) {
-	path := os.Getenv("SHELL")
-	if path == "" {
-		return nil, errors.New("unable to detect the current shell")
+// NewShell initializes the Shell struct so it can be used to start a shell environment
+// for the devbox project.
+func NewShell(opts ...ShellOption) (*Shell, error) {
```

Comment:
What about `UserShell` to differentiate it from a Nix shell which is always `bash`? I wanted to avoid `New` because it doesn't actually create a shell yet.

Reply:
yeah, the problem is that the word "shell" is overloaded here, because it could refer to both the struct `Shell` and the underlying system shell. When I write `NewShell`, I am using it in the sense of `New{StructName}` as a struct constructor.

To mitigate this confusion, would it make sense to rename the struct to `type DevboxShell struct`? This function would then become `NewDevboxShell`

Reply:
One reason I intentionally _do_ want to use the word `New` is to make it clear that we have this constructor function, so that callers do not do `nix.Shell{}` and feel they have a struct that is usable.

Reply:
> yeah, the problem is that the word "shell" is overloaded here, because it could refer to both the struct Shell and the underlying system shell.

The original thought was to have the struct represent a shell, in the same way that `exec.Cmd` does. You call `DetectShell` and you get a shell that you can run or inspect. If you'd like to use `New`, I'd prefer if we could still keep it descriptive to what it's doing -- guessing the user's shell vs. using a default. `NewUserShell` or `NewDefaultShell`, etc.

> To mitigate this confusion, would it make sense to rename the struct to type DevboxShell struct

I agree that the name doesn't really make sense anymore. It was originally intended to be just a Nix shell, with the Devbox-specific logic going in the `devbox` package. IMO the best option would be to just move it out of the Nix package since it doesn't have anything to do with Nix anymore. I know what's a larger undertaking though, so just a rename to `DevboxShell` would be okay with me.

> One reason I intentionally do want to use the word New is to make it clear that we have this constructor function, so that callers do not do nix.Shell{} and feel they have a struct that is usable.

The zero value of `Shell` is usable - it starts a bash Nix shell. The original intention was for a `nix.Shell{}` to be usable and the caller could further modify the fields they needed to give it a different environment or init script. `DetectShell` was a convenience to go through the work of configuring those fields to point to the user's preferred shell and init files. I'd still prefer to keep the zero-value working as a default Nix/bash shell if possible. This is especially useful for tests.

Reply:
I've made the following updates:
1. Rename `nix.Shell` to `nix.DevboxShell`.
2. Added a TODO to move `DevboxShell` to `impl` package
3. Renamed `NewShell` to `NewDevboxShell`.

In a future PR, one of us can refactor out code to solve the TODO of (2) and also split it so that there's once again a rudimentary `nix.Shell` struct as before. 

I'm not convinced that its a good idea to have a usable raw struct, since it leads to the kinds of problems this PR is solving. Bugs like [this line](https://github.com/jetpack-io/devbox/blob/main/internal/impl/devbox.go#L277) being introduced which lose all the other shell options that should be applied to it. For tests, one can have a `NewTestShell` or similar constructor function if needed.

Reply:
That wasn't really a bug though. The intended behavior was to start a default Nix shell if we couldn't figure out the shell type. You can't set any of the shell options unless you know which flags and environment variables the specific shell uses. For example, how do you set the PATH if you don't know how to point the shell to the devbox rcfile?

Reply:
Maybe I was not quite understanding the desired behavior then. To me, it seemed better to fail if we can't resolve the shell such that we can apply all the plugins and devbox-project settings appropriately...which is why this PR errors out. I think the new logic to detect the nix-bash-shell does make that error scenario pretty unlikely though.